Enhance L7 NetworkPolicy to support TLS protocol (#4932)
antrea-io/antrea: docs/antrea-l7-network-policy.md

- [Prerequisites](#prerequisites)
- [Usage](#usage)
  - [HTTP](#http)

>     - [More examples - HTTP](#more-examples-http)

Reviewer: This is inconsistent with the actual section name. TOC check should be failing.
And I feel redundant to append HTTP when its parent section is already HTTP. If you are trying to avoid section name conflict, the second anchor of the second section will have "-1", and toc generator tool will do it automatically.

Author: I didn't know that we could generate TOC by calling `make toc`. I have updated and checked.

### HTTP

> An example of layer 7 NetworkPolicy for the HTTP protocol is like below:

Reviewer 2: Remove "of"

Author: Done, thanks for reviewing this PRs.

        - http: {}       # automatically dropped, and subsequent rules will not be considered.
```

### TLS

> A typical layer 7 NetworkPolicy TLS protocol is as below:

Reviewer: ```suggestion
A typical layer 7 NetworkPolicy for TLS protocol is as below:
```

Author: Done

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: NetworkPolicy
metadata:

>   name: ingress-allow-tls-request-to-api-v2

Reviewer: The name doesn't reflect what the policy really does

Author: Updated to `ingress-allow-tls-hanshake`

spec:
  priority: 5
  tier: application
  appliedTo:
    - podSelector:
        matchLabels:
          app: web
  ingress:

>     - name: allow-http   # Allow inbound TLS requests to TLS server name "foo.bar.com" from Pods with app=client label.

Reviewer: ditto

Author: Done

      action: Allow      # All other traffic from these Pods will be automatically dropped, and subsequent rules will not be considered.
      from:
        - podSelector:
            matchLabels:
              app: client
      l7Protocols:
        - tls:
            sni: "foo.bar.com"
    - name: drop-other   # Drop all other inbound traffic (i.e., from Pods without the app=client label or from external clients).
      action: Drop
```

> **sni**: The `sni` field represents the service name present in TLS header to match. Both exact matches and wildcards are

Reviewer: ```suggestion
**sni**: The `sni` field matches the TLS/SSL Server Name Indication (SNI) field included in the TLS/SSL handshake process. Both exact matches and wildcards are
```

Author: Done

> supported, e.g. `*.foo.com`, `*.foo.*`, `foo.bar.com`. If not set, the rule matches any service name.

Reviewer: ```suggestion
supported, e.g. `*.foo.com`, `*.foo.*`, `foo.bar.com`. If not set, the rule matches all names.
```

Author: Done

#### More examples TLS

The following NetworkPolicy prevents applications from accessing unauthorized TLS server names:

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ClusterNetworkPolicy
metadata:

>   name: allow-web-tls-access-to-internal-domain

Reviewer: web should not be part of it. this rule doesn't indicate web at all.

Author: Updated to ` allow-tls-handshake-to-internal`

spec:
  priority: 5
  tier: securityops
  appliedTo:
    - podSelector:
        matchLabels:
          egress-restriction: internal-domain-only
  egress:
    - name: allow-dns          # Allow outbound DNS requests.
      action: Allow
      ports:
        - protocol: TCP
          port: 53
        - protocol: UDP
          port: 53

>     - name: allow-tls-only    # Allow outbound TLS requests towards foo.bar.com. 

Reviewer: this is inconsistent with the actual server name

      action: Allow           # As the rule's "to" and "ports" are empty, which means it selects traffic to any network
      l7Protocols:            # peer's any port using any transport protocol, all outbound TLS requests towards other

>         - tls:                # service names and non-TLS requests will be automatically dropped, and subsequent rules

Reviewer: not sure why you use "service name" in several places. It should always be "server name".

Author: My fault. It should be "server name".

            sni: "*.bar.com"  # will not be considered.
```

The following NetworkPolicy blocks network traffic using an unauthorized application protocol regardless of port used.

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: NetworkPolicy
metadata:

>   name: allow-http-only

Reviewer: wrong name

Author: Updated to `allow-tls-only`

        - http: {}       # automatically dropped, and subsequent rules will not be considered.
```

### TLS

> A typical layer 7 NetworkPolicy for TLS protocol is as below:

Reviewer 2: A typical -> An example

Reviewer 2: as -> like

Reviewer 2: the TLS protocol

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: NetworkPolicy
metadata:
  name: ingress-allow-tls-handshake
spec:
  priority: 5
  tier: application
  appliedTo:
    - podSelector:
        matchLabels:
          app: web
  ingress:

>     - name: allow-tls    # Allow inbound TLS/SSL handshakes to server name "foo.bar.com" from Pods with app=client label.

Reviewer 2: "handshake requests" or "handshake packets"

Reviewer 2: with label "app=client"

      action: Allow      # All other traffic from these Pods will be automatically dropped, and subsequent rules will not be considered.
      from:
        - podSelector:
            matchLabels:
              app: client
      l7Protocols:
        - tls:
            sni: "foo.bar.com"
    - name: drop-other   # Drop all other inbound traffic (i.e., from Pods without the app=client label or from external clients).
      action: Drop
```

> **sni**: The `sni` field matches the TLS/SSL Server Name Indication (SNI) field included in the TLS/SSL handshake process.

Reviewer 2: Remove "included"?

Both exact matches and wildcards are supported, e.g. `*.foo.com`, `*.foo.*`, `foo.bar.com`. If not set, the rule matches
all names.

#### More examples

The following NetworkPolicy prevents applications from accessing unauthorized SSL/TLS server names:

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ClusterNetworkPolicy
metadata:
  name: allow-tls-handshake-to-internal
spec:
  priority: 5
  tier: securityops
  appliedTo:
    - podSelector:
        matchLabels:
          egress-restriction: internal-tls-only
  egress:
    - name: allow-dns          # Allow outbound DNS requests.
      action: Allow
      ports:
        - protocol: TCP
          port: 53
        - protocol: UDP
          port: 53
    - name: allow-tls-only      # Allow outbound SSL/TLS handshakes towards "*.bar.com". 
      action: Allow             # As the rule's "to" and "ports" are empty, which means it selects traffic to any network

>       l7Protocols:              # peer's any port using any transport protocol, all outbound SSL/TLS handshakes towards other

Reviewer 2: using -> of

        - tls:                  # server names and non-SSL/non-TLS handshakes will be automatically dropped, and subsequent rules
            sni: "*.bar.com"    # will not be considered.
```

> The following NetworkPolicy blocks network traffic using an unauthorized application protocol regardless of port used.

Reviewer 2: port -> the port

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: NetworkPolicy
metadata:
  name: allow-tls-only
spec:
  priority: 5
  tier: application
  appliedTo:
    - podSelector:
        matchLabels:
          app: web
  ingress:

>     - name: tls-only     # Allow inbound SSL/TLS requests only.

Reviewer 2: handshake requests
